designs/2022-parsing-session-objects/README.md

```diff
@@ -0,0 +1,247 @@
+- Repo: eslint/eslint
+- Start Date: 2022-12-30
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: [Brad Zacher](https://github.com/bradzacher), [Josh Goldberg](https://github.com/JoshuaKGoldberg)
+
+# Parsing Session Objects
+
+## Summary
+
+Parsers are not currently told whether ESLint is being run in a CLI-style single pass, a single pass with `--fix` enabled, or a long-running IDE session.
+This RFC proposes the creation of a "session" object that provides them this information.
+
+## Motivation
+
+Forking from https://github.com/eslint/eslint/discussions/16557#discussioncomment-4219483: custom parsers sometimes would want to make decisions in stateful logic based on the way ESLint is being run.
+There are generally three classifications ESLint sessions for these stateful parsers:
+
+- "Single" runs: which can have a reusable (_immutable_ program) cache, and doesn't need to attach file watchers
+- "Single, with fixers": which have a _mostly_ reusable cache, except for file fixes, which we generally assume not to change cross-file types
+- "Continuous" runs: the "wild west" (_builder_ program), and should attach file watchers
+
+From knowing session classification, parsers can then make more informed decision on aspects such as file/program caching.
+
+As discussed in https://github.com/eslint/eslint/discussions/16557#discussioncomment-4160219, at least two major ecosystem plugins would benefit from understanding session information at _parse_ time:
+
+- [`eslint-plugin-import`](https://github.com/import-js/eslint-plugin-import)'s imports analysis
```

Comment:
@ljharb do you want to add anything about eslint-plugin-import? I think it'd help the RFC to know whether & how session objects would be used there. Especially around your [suggestion of making them unchanged for the duration of a run](https://github.com/eslint/eslint/issues/16656#issuecomment-1347048192).

Reply:
Specifically, I'd use them as a weak collection key in all my caches - that way, everything would be maximally cached per session, but the caches wouldn't go stale in a long-running eslint process.

Reply:
Thanks - added a section in 6b7f885. I haven't had the time to deep dive into `eslint-plugin-import`'s parsing mechanics to write anything deeper. If you want to add some more thorough evidence for the proposal, I'd be happy to add you as a co-author. 😄